step/step.go

```diff
@@ -0,0 +1,113 @@
+package step
+
+import (
+	"fmt"
+
+	"github.com/bitrise-io/go-steputils/v2/stepconf"
+	"github.com/bitrise-io/go-utils/v2/command"
+	"github.com/bitrise-io/go-utils/v2/log"
+	"github.com/bitrise-steplib/steps-git-clone/gitclone"
+)
+
+type Input struct {
+	ShouldMergePR bool `env:"merge_pr,opt[yes,no]"`
+
+	CloneIntoDir         string   `env:"clone_into_dir,required"`
+	CloneDepth           int      `env:"clone_depth"`
+	UpdateSubmodules     bool     `env:"update_submodules,opt[yes,no]"`
+	SubmoduleUpdateDepth int      `env:"submodule_update_depth"`
+	FetchTags            bool     `env:"fetch_tags,opt[yes,no]"`
+	SparseDirectories    []string `env:"sparse_directories,multiline"`
+
+	RepositoryURL         string `env:"repository_url,required"`
+	Commit                string `env:"commit"`
+	Tag                   string `env:"tag"`
+	Branch                string `env:"branch"`
+	PRDestBranch          string `env:"branch_dest"`
+	PRSourceRepositoryURL string `env:"pull_request_repository_url"`
+	PRMergeBranch         string `env:"pull_request_merge_branch"`
+	PRHeadBranch          string `env:"pull_request_head_branch"`
+
+	ResetRepository bool   `env:"reset_repository,opt[Yes,No]"`
+	BuildURL        string `env:"build_url"`
+	BuildAPIToken   string `env:"build_api_token"`
+}
+
+// Config is the git clone step configuration
+type Config struct {
+	Input
+}
+
+type GitCloneStep struct {
+	logger      log.Logger
+	tracker     gitclone.StepTracker
+	inputParser stepconf.InputParser
+	cmdFactory  command.Factory
+}
+
+func NewGitCloneStep(logger log.Logger, tracker gitclone.StepTracker, inputParser stepconf.InputParser, cmdFactory command.Factory) GitCloneStep {
+	return GitCloneStep{
+		logger:      logger,
+		tracker:     tracker,
+		inputParser: inputParser,
+		cmdFactory:  cmdFactory,
+	}
+}
+
+func (g GitCloneStep) ProcessConfig() (Config, error) {
+	var cfg Config
+	if err := g.inputParser.Parse(&cfg); err != nil {
+		return Config{}, fmt.Errorf("Error: %s\n", err)
+	}
+	stepconf.Print(cfg)
+
+	return cfg, nil
+}
+
+func (g GitCloneStep) Run(cfg Config) (gitclone.CheckoutStateResult, error) {
+	gitcloneCfg := convertConfig(cfg)
+	cloner := gitclone.NewGitCloner(g.logger, g.tracker, g.cmdFactory)
+	return cloner.CheckoutState(gitcloneCfg)
+}
+
+func (g GitCloneStep) ExportOutputs(runResult gitclone.CheckoutStateResult) error {
```

Comment:
This `CheckoutStateResult` seems a bit smelly to me. If my understanding is correct, you introduced it in order to expose a public type from the `gitclone` package so that we can pass the info between `Run()` and `ExportOutputs()` here. I see a couple of problems with it:

1. A few internal types had to be converted to public because they are now part of the "result" and `step.go` needs to access them.
2. The `GitCmd` within the result is especially smelly, since it's not even a result, but I see that in line 84 it is used in a constructor. This feels like unnecessary ping-pong between the `gitclone` and this `step` package.

Can we maybe avoid "uplifting" these types to the `step` package and let the `gitclone` package handle them somehow? I feel like that would be cleaner and involve less boilerplate code.

Reply:
Please check these commits:
8d881cb
94bb2a7

Reply:
LGTM now!